testonly/map_metadata.go

```diff
@@ -0,0 +1,35 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testonly
+
+import(
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+)
+
+// MustMarshalMetadata is used in tests to Marshal proto messages used as map
+// metadata into the protobuf.ptypes.any.Any used in the Trillian API and in
+// storage.  Failure to marshal will fail the test.
+func MustMarshalMetadata(t *testing.T, in proto.Message) *any.Any {
+	var metaAny *any.Any
+	metaAny, err := ptypes.MarshalAny(in)
+	if err != nil {
+		t.Fatalf("failed to marshal %v as 'any': err %v", in, err)
```

Comment:
FYI, our other must*Something* functions tend to have a single parameter and panic on failures. Both are acceptable, but I've grown to slightly prefer the panic version, FWIW.

Reply:
Won't `panic()`ing here cause the whole test suite to fail?

Reply:
In the context of a "MustMarshal" method that's fine. The assumption behind MustMarshal calls is that the test is working with canned, known-to-be-valid data. If the marshal breaks something is very wrong. It's "never should happen" territory.

Reply:
I don't agree - if there multiple things wrong then you're masking the others by panic()ing; I think it's better to have a complete picture of what needs to be fixed rather than serially picking your way through multiple failures.

Reply:
I think you're picking the wrong bone. Have a look at mustMarshalAny usages: they all start with hand-written, valid protos. If the proto is wrong, the entire test is broken. You wouldn't get much more out of Fatalf either, if anything.

It bears repeating: this is for hand-written, known-to-be-valid protos used in tests. Even if wanted to, I find it difficult to conjure a hand-written proto that breaks the marshaling.